File: platform/logging-service/src/LoggingService.Core/Models/LogEntry.cs
using System.Text.Json.Serialization;

namespace LoggingService.Core.Models;

/// <summary>
/// Represents a structured log entry from a microservice.
/// </summary>
public sealed record LogEntry
{
    /// <summary>
    /// Unique identifier for the log entry.
    /// </summary>
    public required string Id { get; init; }

    /// <summary>
    /// Timestamp when the log was generated (ISO 8601 UTC).
    /// </summary>
    public required DateTimeOffset Timestamp { get; init; }

    /// <summary>
    /// Correlation ID for tracing requests across services.
    /// </summary>
    public required string CorrelationId { get; init; }

    /// <summary>
    /// Identifier of the service that generated the log.
    /// </summary>
    public required string ServiceId { get; init; }

    /// <summary>
    /// Severity level of the log entry.
    /// </summary>
    public required LogLevel Level { get; init; }

    /// <summary>
    /// Log message content.
    /// </summary>
    public required string Message { get; init; }

    /// <summary>
    /// OpenTelemetry trace ID for distributed tracing.
    /// </summary>
    public string? TraceId { get; init; }

    /// <summary>
    /// OpenTelemetry span ID for distributed tracing.
    /// </summary>
    public string? SpanId { get; init; }

    /// <summary>
    /// User identifier associated with the log entry.
    /// </summary>
    public string? UserId { get; init; }

    /// <summary>
    /// Request identifier for the originating request.
    /// </summary>
    public string? RequestId { get; init; }

    /// <summary>
    /// HTTP method of the request (GET, POST, etc.).
    /// </summary>
    public string? Method { get; init; }

    /// <summary>
    /// Request path or endpoint.
    /// </summary>
    public string? Path { get; init; }

    /// <summary>
    /// HTTP status code of the response.
    /// </summary>
    public int? StatusCode { get; init; }

    /// <summary>
    /// Duration of the operation in milliseconds.
    /// </summary>
    public long? DurationMs { get; init; }

    /// <summary>
    /// Additional metadata as key-value pairs.
    /// </summary>
    public Dictionary<string, object>? Metadata { get; init; }

    /// <summary>
    /// Exception information if the log is related to an error.
    /// </summary>
    public ExceptionInfo? Exception { get; init; }
}

/// <summary>
/// Log severity levels.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter<LogLevel>))]
public enum LogLevel
{
    Debug = 0,
    Info = 1,
    Warn = 2,
    Error = 3,
    Fatal = 4
}

/// <summary>
/// Exception information for error logs.
/// </summary>
public sealed record ExceptionInfo
{
    /// <summary>
    /// Exception type name.
    /// </summary>
    public required string Type { get; init; }

    /// <summary>
    /// Exception message.
    /// </summary>
    public required string Message { get; init; }

    /// <summary>
    /// Stack trace of the exception.
    /// </summary>
    public string? StackTrace { get; init; }

    /// <summary>
    /// Inner exception information.
    /// </summary>
    public ExceptionInfo? InnerException { get; init; }
}

File: services/email-service/src/Infrastructure/Platform/LogEntry.php
<?php

declare(strict_types=1);

namespace EmailService\Infrastructure\Platform;

/**
 * Structured log entry for platform logging-service.
 */
final readonly class LogEntry
{
    public function __construct(
        public string $correlationId,
        public string $serviceId,
        public LogLevel $level,
        public string $message,
        public ?string $traceId = null,
        public ?string $spanId = null,
        public ?string $userId = null,
        public ?string $requestId = null,
        public ?string $method = null,
        public ?string $path = null,
        public ?int $statusCode = null,
        public ?int $durationMs = null,
        public array $metadata = [],
        public ?ExceptionInfo $exception = null,
        public ?\DateTimeImmutable $timestamp = null,
    ) {}

    public static function info(string $correlationId, string $message, array $metadata = []): self
    {
        return new self(
            correlationId: $correlationId,
            serviceId: 'email-service',
            level: LogLevel::INFO,
            message: $message,
            metadata: $metadata,
            timestamp: new \DateTimeImmutable(),
        );
    }

    public static function error(
        string $correlationId,
        string $message,
        ?\Throwable $exception = null,
        array $metadata = [],
    ): self {
        return new self(
            correlationId: $correlationId,
            serviceId: 'email-service',
            level: LogLevel::ERROR,
            message: $message,
            metadata: $metadata,
            exception: $exception !== null ? ExceptionInfo::fromThrowable($exception) : null,
            timestamp: new \DateTimeImmutable(),
        );
    }

    public static function warn(string $correlationId, string $message, array $metadata = []): self
    {
        return new self(
            correlationId: $correlationId,
            serviceId: 'email-service',
            level: LogLevel::WARN,
            message: $message,
            metadata: $metadata,
            timestamp: new \DateTimeImmutable(),
        );
    }

    public static function debug(string $correlationId, string $message, array $metadata = []): self
    {
        return new self(
            correlationId: $correlationId,
            serviceId: 'email-service',
            level: LogLevel::DEBUG,
            message: $message,
            metadata: $metadata,
            timestamp: new \DateTimeImmutable(),
        );
    }

    public function toArray(): array
    {
        $data = [
            'correlation_id' => $this->correlationId,
            'service_id' => $this->serviceId,
            'level' => $this->level->value,
            'message' => $this->message,
            'timestamp' => ($this->timestamp ?? new \DateTimeImmutable())->format(\DateTimeInterface::RFC3339_EXTENDED),
        ];

        if ($this->traceId !== null) {
            $data['trace_id'] = $this->traceId;
        }
        if ($this->spanId !== null) {
            $data['span_id'] = $this->spanId;
        }
        if ($this->userId !== null) {
            $data['user_id'] = $this->userId;
        }
        if ($this->requestId !== null) {
            $data['request_id'] = $this->requestId;
        }
        if ($this->method !== null) {
            $data['method'] = $this->method;
        }
        if ($this->path !== null) {
            $data['path'] = $this->path;
        }
        if ($this->statusCode !== null) {
            $data['status_code'] = $this->statusCode;
        }
        if ($this->durationMs !== null) {
            $data['duration_ms'] = $this->durationMs;
        }
        if (!empty($this->metadata)) {
            $data['metadata'] = $this->metadata;
        }
        if ($this->exception !== null) {
            $data['exception'] = $this->exception->toArray();
        }

        return $data;
    }
}
